src/testflinger_device_connectors/devices/oemrecovery/oemrecovery.py

```diff
@@ -89,6 +91,68 @@ def provision(self):
         self._run_cmd_list(recovery_cmds)
         self.check_device_booted()
 
+    @timeout(dec_timeout=3600)
+    def check_boot_state(self):
+        """check boot state"""
+        try:
+            test_username = self.job_data.get("test_data", {}).get(
+                "test_username", "ubuntu"
+            )
+            test_password = self.job_data.get("test_data", {}).get(
+                "test_password", "ubuntu"
+            )
+        except AttributeError:
+            test_username = "ubuntu"
+            test_password = "ubuntu"
+
+        while True:
+            try:
+                image_type = (
+                    subprocess.check_output(
+                        (
+                            "sshpass -p {} ssh -o StrictHostKeyChecking=no "
+                            "-o UserKnownHostsFile=/dev/null "
+                            "{}@{} snap model | grep model"
+                        ).format(
+                            test_password,
+                            test_username,
+                            self.config["device_ip"],
+                        ),
+                        shell=True,
+                    )
+                ).decode()
+            except Exception:
+                time.sleep(10)
+                continue
+
+            if image_type.find("generic-classic") != -1:
+                return
+
+            else:
+                while True:
+                    try:
+                        boot_state = (
+                            subprocess.check_output(
+                                (
+                                    "sshpass -p {} ssh "
+                                    "-o StrictHostKeyChecking=no "
+                                    "-o UserKnownHostsFile=/dev/null "
+                                    "{}@{} snap changes "
+                                    '| grep "Initialize device"'
+                                ).format(
+                                    test_password,
+                                    test_username,
+                                    self.config["device_ip"],
+                                ),
+                                shell=True,
+                            )
+                        ).decode()
+
+                        if boot_state != "" and boot_state.find("Done") != -1:
```

Comment:
You can replace all of this with `if "Done" in boot_state` I think.

Reply:
Thank you, another lesson to me